File: examples/inverse-dynamics.cpp
// Copyright 2023 Inria
// SPDX-License-Identifier: BSD-2-Clause

// ============================================================
// 逆动力学（Inverse Dynamics）：RNEA 计算关节力矩
//
// 物理含义：给定运动学轨迹 (q, v, a)，计算驱动该运动所需的关节力矩 τ
// 应用场景：
//   - 前馈力矩计算（轨迹跟踪控制器的前馈项）
//   - 系统辨识（用已知运动反求惯量参数）
//   - 重力补偿（v=0, a=0 时 τ = g(q)）
//
// RNEA（Recursive Newton-Euler Algorithm）：
//   正向传播：从根到末端计算各连杆的速度和加速度
//   反向传播：从末端到根计算各连杆的关节力矩
//   时间复杂度：O(n)
//
// 完整动力学方程：τ = M(q)·a + C(q,v)·v + g(q)
//   RNEA 一次调用直接给出总力矩，不分别计算 M, C, g
// ============================================================

#include <iostream>

#include "pinocchio/algorithm/joint-configuration.hpp"  // randomConfiguration
#include "pinocchio/algorithm/rnea.hpp"                 // rnea
#include "pinocchio/parsers/urdf.hpp"                   // urdf::buildModel

// PINOCCHIO_MODEL_DIR is defined by the CMake but you can define your own
// directory here.
#ifndef PINOCCHIO_MODEL_DIR
  #define PINOCCHIO_MODEL_DIR "path_to_the_model_dir"
#endif

int main(int argc, char ** argv)
{
  using namespace pinocchio;

  // Change to your own URDF file here, or give a path as command-line argument
  const std::string urdf_filename = (argc <= 1)
                                      ? PINOCCHIO_MODEL_DIR
                                          + std::string("/example-robot-data/robots/"
                                                        "ur_description/urdf/ur5_robot.urdf")
                                      : argv[1];

  // Load the URDF model
  Model model;
  pinocchio::urdf::buildModel(urdf_filename, model);

  // Build a data frame associated with the model
  Data data(model);

  // Sample a random joint configuration, joint velocities and accelerations
  Eigen::VectorXd q = randomConfiguration(model);      // in rad for the UR5
  Eigen::VectorXd v = Eigen::VectorXd::Zero(model.nv); // in rad/s for the UR5（v=0：静止）
  Eigen::VectorXd a = Eigen::VectorXd::Zero(model.nv); // in rad/s² for the UR5（a=0：重力补偿）

  // ---- RNEA（逆动力学）----
  // 当 v=0, a=0 时：rnea 返回重力补偿力矩 g(q)
  // 结果同时存储在 data.tau（返回值是对 data.tau 的 const 引用）
  Eigen::VectorXd tau = pinocchio::rnea(model, data, q, v, a);

  // Print out to the vector of joint torques (in N.m)
  // data.tau 和 tau 指向同一内存（tau 是 data.tau 的引用拷贝）
  std::cout << "Joint torques: " << data.tau.transpose() << std::endl;
  return 0;
}

File: examples/inverse-dynamics.py
# Copyright 2023 Inria
# SPDX-License-Identifier: BSD-2-Clause


"""
In this short script, we show how to compute inverse dynamics (RNEA), i.e. the
vector of joint torques corresponding to a given motion.
"""

from pathlib import Path

import numpy as np
import pinocchio as pin

# 模型目录，替换为你自己的 URDF 路径
pinocchio_model_dir = Path(__file__).parent.parent / "models/"
model_path = pinocchio_model_dir / "example-robot-data/robots"
mesh_dir = pinocchio_model_dir
urdf_filename = "ur5_robot.urdf"
urdf_model_path = model_path / "ur_description/urdf/" / urdf_filename

# buildModelsFromUrdf 同时返回三个模型：
#   model:           运动学/动力学模型（关节拓扑 + 惯量参数）
#   collision_model: 碰撞几何模型（用于碰撞检测，基于 hpp-fcl）
#   visual_model:    视觉几何模型（用于可视化渲染）
# 这里只用运动学模型，忽略另外两个
model, _, _ = pin.buildModelsFromUrdf(urdf_model_path, package_dirs=mesh_dir)

# 创建算法数据容器
data = model.createData()

# 随机采样关节状态（实际使用时替换为传感器读数）
q = pin.randomConfiguration(model)  # 关节角 [rad]，满足关节限位约束
v = np.random.rand(model.nv, 1)     # 关节速度 [rad/s]
a = np.random.rand(model.nv, 1)     # 期望关节加速度 [rad/s²]

# RNEA（递推牛顿-欧拉算法）：逆动力学
# 给定 (q, v, a)，计算产生该运动所需的关节力矩 τ
# 等价方程：τ = M(q)·a + C(q,v)·v + g(q)
# 算法复杂度：O(n)，远优于显式构造 M, C, g 矩阵再相乘（O(n²)~O(n³)）
# 结果同时存储在返回值和 data.tau 中
tau = pin.rnea(model, data, q, v, a)

# 打印关节力矩（单位：N·m）
print("Joint torques: " + str(tau))
